### PR TITLE
fix(core): check if transplanted views are attached to change detector

### PR DESCRIPTION
**Pull request by edusperoni**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/main/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:

## What is the current behavior?
Template views are always changed detected inside OnPush components if they are detached. Depending on the project, it needs to have at least 1 transplanted view attached for the issue to be reproduced.

Issue Number: #46973

## What is the new behavior?

We first check if the transplanted view is attached to change detection before running any other CD in it.

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: packages/core/test/acceptance/change_detection_transplanted_view_spec.ts [hunk 1]
```diff
@@ -475,10 +475,17 @@ describe('change detection for transplanted views', () => {
   describe('transplanted views shielded by OnPush', () => {
     @Component({
       selector: 'check-always-insertion',
-      template: `<ng-container [ngTemplateOutlet]="template"></ng-container>`
+      template:
+          `<ng-container [ngTemplateOutlet]="template"></ng-container><ng-container #dynamic></ng-container>`
```
[the remark this answers is not in the document]
Comment (atscott) [reply packages/core/test/acceptance/change_detection_transplanted_view_spec.ts:479]: Or remove based on comment below

File: packages/core/test/acceptance/change_detection_transplanted_view_spec.ts [hunk 2]
```diff
@@ -475,10 +475,17 @@ describe('change detection for transplanted views', () => {
...
     })
     class CheckAlwaysInsertion {
       @Input() template!: TemplateRef<{}>;
+      @ViewChild('dynamic', {read: ViewContainerRef}) detached!: ViewContainerRef;
+
+      ngAfterViewInit() {
+        // create a detached template. This template should never be change detected.
+        this.detached.createEmbeddedView(this.template).detach();
+      }
```
Comment (atscott) [inline packages/core/test/acceptance/change_detection_transplanted_view_spec.ts:488]: I realize the test cases implicitly test this because the inner text is never "new new" but it would be much better to make this more explicit. 

Can we update the tests to have an attach/detach mechanism that allows us to directly test the behavior of skipping detached views? In this case, I don't think you'd really need a second container but could instead just update the existing oner to be attach/detachable.

We should also cover the case where the view is re-attached.

**Review by atscott — CHANGES_REQUESTED** [review]
Changes look good, pending verification via internal presubmit checks. I've requested some changes to the tests.

File: packages/core/test/acceptance/change_detection_transplanted_view_spec.ts (hunk 2 above)
Comment (edusperoni) [reply packages/core/test/acceptance/change_detection_transplanted_view_spec.ts:488]: Thank you for the feedback! I agree with you on the test scenarios. When I originally wrote it I thought making the view attachable/detachable would pollute the tests too much.

Any thoughts on how we could get the instance to the EmbeddedViewRef from the ng-template?

Also, there's one specific case where this only breaks when there are 2 templates inside (at least 1 attached), as in some scenarios having 1 attached view will make the view count be > 1 and the HAS_TRANSPLANTED_VIEWS be true, which will make it CD all the inner templates (even detached ones). So having only 1 detached template would not be enough to cover all cases

Comment (atscott) [reply packages/core/test/acceptance/change_detection_transplanted_view_spec.ts:488]: > Also, there's one specific case where this only breaks when there are 2 templates inside (at least 1 attached), as in some scenarios having 1 attached view will make the view count be > 1 and the HAS_TRANSPLANTED_VIEWS be true, which will make it CD all the inner templates (even detached ones). So having only 1 detached template would not be enough to cover all cases

Ah, right. I forgot about that again :) 

> Any thoughts on how we could get the instance to the EmbeddedViewRef from the ng-template?

I'm not sure I follow the question here. `createEmbeddedView` gives you the `EmbeddedViewRef` that you're calling `detach` on. You can save that ref to a public class member and call `attach/detach` on it at will.

Lastly, after reviewing the tests again, I think this should really be another `describe` block altogether. Putting it with the set of tests under "transplanted views shielded by OnPush" is actually putting it somewhere unrelated to what we've changed. The `OnPush` parent with a `CheckAlways` child as the insertion isn't necessary. All you really need to exercise the behavior is a single `OnPush` insertion.

**Comment by atscott** [comment]
All internal tests passed in the [TGP](https://fusion2.corp.google.com/presubmit/tap/464574397/OCL:464574397:BASE:464543952:1659392273198:10643d5/targets). This change is good to go once comments are resolved.

**Comment by edusperoni** [comment]
@atscott I've updated the PR with new tests. They should cover all cases now

**Review by atscott — APPROVED** [review]
Perfect 👌

**Comment by dylhunn** [comment]
This PR was merged into the repository by commit dbed2cf079c52825ce0aa67fc2a46bc3d38f408f.
